src/algebra/module/pid.lean

```diff
@@ -0,0 +1,288 @@
+/-
+Copyright (c) 2022 Pierre-Alexandre Bazin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pierre-Alexandre Bazin
+-/
+import algebra.module.torsion
+import linear_algebra.free_module.pid
+import algebra.module.projective
+import algebra.category.Module.biproducts
+import ring_theory.dedekind_domain.ideal
+
+/-!
+# Structure of finitely generated modules over a PID
+
+## Main statements
+
+* `equiv_direct_sum_of_is_torsion` : A finitely generated torsion module over a PID is isomorphic
+  to a direct sum of some `R ⧸ R ∙ (p i ^ e i)` where the `p i ^ e i` are prime powers.
+* `equiv_free_prod_direct_sum` : A finitely generated module over a PID is isomorphic to the
+  product of a free module (its torsion free part) and a direct sum of the form above (its torsion
+  submodule).
+
+## Notation
+
+* `R` is a PID and `M` is a (finitely generated for main statements) torsion module over R.
+* `p` is an irreducible element of `R` or a tuple of these.
+
+## Implementation details
+
+We first prove (`is_internal_prime_power_torsion`) that a finitely generated torsion module is the
+internal direct sum of its `p i ^ e i`-torsion submodules for some (finitely many) prime powers
+`p i ^ e i`.
+
+Then we treat the case of a `p ^ ∞`-torsion module (that is, a module where all elements are
+cancelled by scalar multiplication by some power of `p`) and apply it to the `p i ^ e i`-torsion
+submodules (that are `p i ^ ∞`-torsion) to get the result for torsion modules.
+
+Then we get the general result using that a torsion free module is free (which has been proved at
+`module.free_of_finite_type_torsion_free'` at `linear_algebra/free_module/pid.lean`.)
+
+## Tags
+
+Finitely generated module, principal ideal domain, classification, structure theorem
+-/
+
+universes u v
+open_locale big_operators
+
+variables {R : Type u} [comm_ring R] [is_domain R] {M : Type v} [add_comm_group M] [module R M]
+variables {N : Type (max u v)} [add_comm_group N] [module R N]
+
+open_locale direct_sum
+open submodule
+
+lemma is_torsion_by_ideal_of_finite_of_is_torsion [module.finite R M] (hM : module.is_torsion R M) :
```

Comment:
Can you move this to a more basic file? I think it belongs to the file where `module.is_torsion` is defined.

Reply:
Done at #14573